### PR TITLE
Fix off by one error in TextPainter.

### LINKED ISSUE #15572: Editing multiline text sometimes places cursor on the wrong line
## Steps to Reproduce
Open the Text Fields demo in Flutter Gallery. Write some text in the Life Story field, hit the return key on the keyboard. The text cursor doesn't move to the next line until you type another character.

```
[✓] Flutter (Channel dev, v0.2.0, on Mac OS X 10.13.2 17C88, locale en-US)
    • Flutter version 0.2.0 at /Users/vik/Projects/flutter
    • Framework revision f8ac23cd86 (10 days ago), 2018-03-05 23:27:57 +0100
    • Engine revision a00f8e8bc0
    • Dart version 2.0.0-dev.32.0.flutter-ee15c8eb68

[!] Android toolchain - develop for Android devices (Android SDK 27.0.3)
    • Android SDK at /Users/vik/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-27, build-tools 27.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-915-b08)
    ✗ Android license status unknown.

[✓] iOS toolchain - develop for iOS devices (Xcode 9.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 9.2, Build version 9C40b
    • ios-deploy 1.9.2
    • CocoaPods version 1.2.0

[✓] Android Studio (version 3.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-915-b08)

[✓] Android Studio (version 2.3)
    • Android Studio at /Applications/Android Studio - old.app/Contents
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] IntelliJ IDEA Community Edition (version 2017.2.5)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 20.0.2
    • Dart plugin version 172.4343.25

[✓] Connected devices (1 available)
    • Viktor’s iPhone • 0f78a8bf61e64931b13636987dc7597b86ba55a7 • ios • iOS 11.2.6

! Doctor found issues in 1 category.
```

### PR DESCRIPTION
**Pull request by jonahwilliams**
Fixes https://github.com/flutter/flutter/issues/15572.

Not sure if this is the exact right solution, but it is close to the problem

Given a string such as `abc`, _getOffsetFromDownstream will return null (out of bounds) and the cursor will correctly be placed at the end of the first line.

Given a string such as `abc\n` _getOffsetFromDownstream will return null and the cursor will incorrectly be placed at the end of the first line.

When deleting text, the affinity changes to upstream and the above issues do not apply.

### CODE REVIEW

**Comment by Hixie** [comment]
Please when writing a test for this make the test comprehensive (check every combination of every possible TextPosition for each of `aaa`, `\naaa`, `a\naa`, `aa\na`, `aaa\n` as well as other esoteric combinations you can come up with like `a\n\na` or `\n\n\n`).

**Comment by jonahwilliams** [comment]
@Hixie I've added testing for the caret position across some different text strings

File: packages/flutter/test/painting/text_painter_test.dart [hunk 1]
```diff
@@ -179,4 +179,59 @@ void main() {
...
+    expect(caretOffset.dy, 42.0);
+
+    /// Moves back to first line when affinity is upstream.
+    text = 'aaa\n';
+    painter.text = new TextSpan(text: text);
+    painter.layout();
+    caretOffset = painter.getOffsetForCaret(new ui.TextPosition(offset: text.length, affinity: TextAffinity.upstream), ui.Rect.zero);
+    expect(caretOffset.dx, painter.width);
+    expect(caretOffset.dy, 0.0);
+  });
```
Comment (Hixie) [inline packages/flutter/test/painting/text_painter_test.dart]: we should check cases with leading newlines too
also for every case we should check both kinds of affinities

Comment (Hixie) [reply packages/flutter/test/painting/text_painter_test.dart]: also we should check positions other than the end of the string. Ideally check every interesting position for each of these strings.

Comment (jonahwilliams) [reply packages/flutter/test/painting/text_painter_test.dart]: I've added cases where I move up and down a text field with leading/trailing newlines with different affinities.  Anything else I may have missed herE?

**Comment by jonahwilliams** [comment]
@Hixie Added more test cases ... I think if we want more we should just fuzz it

**Comment by jonahwilliams** [comment]
@Hixie  friendly ping

**Comment by Hixie** [comment]
LGTM

**Pull request merged**